core/src/main/java/org/infinispan/transaction/xa/recovery/RecoveryManagerImpl.java

```diff
@@ -305,26 +312,29 @@ public String forceTransactionCompletion(Xid xid, boolean commit) {
       }
    }
 
-   private String completeTransaction(LocalTransaction localTx, boolean commit, Xid xid) {
+   private CompletionStage<String> completeTransaction(LocalTransaction localTx, boolean commit, Xid xid) {
+      CompletionStage<?> stage;
+      Function<Throwable, String> errorTranslation;
```

Comment:
IMO this is too abstract, it wouldn't be much longer if everything was inlined:

```
      GlobalTransaction gtx = localTx.getGlobalTransaction();
      if (commit) {
         localTx.clearLookedUpEntries();
         return txCoordinator.prepare(localTx, true)
                             .thenCompose(ignore -> txCoordinator.commit(localTx, false))
                             .thenCompose(ignore -> removeRecoveryInformation(null, xid, gtx, false))
                             .thenApply(ignore -> "Commit successful!")
                             .exceptionally(t -> {
                                log.warnCouldNotCommitLocalTx(localTx, t);
                                return "Could not commit transaction " + xid + " : " + t.getMessage();
                             });
      } else {
         return txCoordinator.rollback(localTx)
                             .thenCompose(ignore -> removeRecoveryInformation(null, xid, gtx, false))
                             .thenApply(ignore -> "Rollback successful")
                             .exceptionally(t -> {
                                log.warnCouldNotRollbackLocalTx(localTx, t);
                                return "Could not rollback transaction " + xid + " : " + t.getMessage();
                             });
      }
```

Reply:
Sure.